usr/src/uts/common/fs/zfs/dmu.c

```diff
@@ -676,9 +736,30 @@ dmu_free_long_range_impl(objset_t *os, dnode_t *dn, uint64_t offset,
 		ASSERT3U(chunk_begin, >=, offset);
 		ASSERT3U(chunk_begin, <=, chunk_end);
 
-		dmu_tx_t *tx = dmu_tx_create(os);
-		dmu_tx_hold_free(tx, dn->dn_object,
-		    chunk_begin, chunk_end - chunk_begin);
+		chunk_len = chunk_end - chunk_begin;
+
+		mutex_enter(&dp->dp_lock);
+		free_dirty_pct = 100 * dp->dp_long_free_dirty_total /
+		    zfs_dirty_data_max;
+		/*
+		 * To avoid filling up a TXG with just frees wait for
+		 * the next TXG to open before freeing more chunks if
+		 * we have reached the threshold of frees for this TXG
+		 */
+		if (dirty_frees_threshold != 0 &&
+		    free_dirty_pct >= dirty_frees_threshold) {
+			mutex_exit(&dp->dp_lock);
+			txg_wait_open(dp, dp->dp_tx.tx_open_txg + 1);
+			continue;
+		}
+		dp->dp_long_free_dirty_total += chunk_len;
+		mutex_exit(&dp->dp_lock);
+
+		DTRACE_PROBE3(free__long__range, uint8_t, free_dirty_pct,
+		    uint64_t, chunk_len, uint64_t, dp->dp_tx.tx_open_txg);
```

Comment:
What is this telling you?  Why not do this after you dmu_tx_assign(), so that you know what txg this chunk will be processed in?